tests/integration/test_pip_packages.py

```diff
@@ -0,0 +1,146 @@
+    assert tarfile.is_tarfile(file_name_tar)
+    with tarfile.open(file_name_tar, "r:gz") as tar:
+        tar.extractall(source_name)
+
+    expected_file_urls = test_env["pip_packages"]["without_deps"]["expected_files"]
+
+    # Check that the source tarball includes the application source code under the app directory.
+    assert_expected_files(path.join(source_name, "app"), expected_file_urls)
+    # Check that the source tarball includes an empty deps directory.
+    assert_expected_files(path.join(source_name, "deps"))
+
+    purl = test_env["pip_packages"]["without_deps"]["purl"]
+    image_contents = [{"dependencies": [], "purl": purl, "sources": []}]
+    assert_pip_content_manifest(client, completed_response.id, image_contents)
+
+
+def assert_packages_from_response(response_data, expected_types):
+    """
+    Check amount and type of packages in the response data.
+
+    :param dict response_data: response data from the Cachito request
+    :param list expected_types: expected types of packages
+    """
+    assert len(response_data["packages"]) == len(expected_types)
+    for idx in range(len(expected_types)):
+        assert response_data["packages"][idx]["type"] == expected_types[idx]
+
+
+def assert_dependencies_from_response(response_data, expected_deps=None):
+    """
+    Check that dependencies are correctly listed in the response data.
+
+    :param dict response_data: response data from the Cachito request
+    :param list expected_deps: expected package dependencies
+    """
+    if expected_deps is None:
+        expected_deps = []
+
+    assert set(response_data["dependencies"]) == set(expected_deps)
+
+    packages_deps = []
+    for package in response_data["packages"]:
+        packages_deps += package["dependencies"]
+    assert set(packages_deps) == set(expected_deps)
+
+
+def assert_expected_files(source_path, expected_file_urls=None):
+    """
+    Check that the source path includes expected files.
+
+    :param str source_path: local path for checking
+    :param dict expected_file_urls: {"relative_path/file_name": "URL", ...}
+    """
+    if expected_file_urls is None:
+        expected_file_urls = {}
+    assert path.exists(source_path) and path.isdir(source_path)
+    assert len(listdir(source_path)) == len(expected_file_urls)
+    files = []
+    # Go through all files in source_code_path and it's subdirectories
+    for root, _, source_files in walk(source_path):
+        for file_name in source_files:
+            # Get path to file in the project
+            absolute_file_path = path.join(root, file_name)
+            relative_file_path = absolute_file_path.split(source_path + "/")[1]
+            file_url = expected_file_urls[relative_file_path]
+            # Download expected file
+            expected_file = requests.get(file_url).text
+            # Assert that content of source file is equal to expected
+            assert open(absolute_file_path, "r").read() == expected_file
```

Comment:
If this function is going to be used to check `.zip` and `.tar.gz` archives, you may need check the binary content instead, i.e. check `response.content` instead of `response.text` and use `"rb"` mode for `open()`

Also, `open()` should be used as a context manager (`with open() as f: ...`)

Reply:
Done!